[BUG] nil pointer dereference in "pkg/clusteragent/clusterchecks".removeConfig

<!--
If you have identified a specific bug in the Agent codebase, please describe it here. If the
issue requires further debugging or investigation, that will be best accomplished via support. Please please contact Datadog [support](http://docs.datadoghq.com/help/) and send them a [flare](https://docs.datadoghq.com/agent/troubleshooting/#send-a-flare) demonstrating the issue.
-->

**Agent Environment**
<!-- please include `agent version` and any other useful output (`agent status`, relevant agent logs, etc.) -->

AWS EKS 1.23
```
# agent version
Cluster Agent 7.44.1 - Commit: 299bdcd - Serialization version: v5.0.76 - Go version: go1.19.8
```

**Describe what happened:**
Agent crashes with the following stack:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x208c53a]

goroutine 517 [running]:
github.com/DataDog/datadog-agent/pkg/clusteragent/clusterchecks.(*dispatcher).removeConfig(0xc001e84080, {0xc00306a3a0, 0x10})
    /go/src/github.com/DataDog/datadog-agent/pkg/clusteragent/clusterchecks/dispatcher_configs.go:99 +0x23a
github.com/DataDog/datadog-agent/pkg/clusteragent/clusterchecks.(*dispatcher).remove(_, {{0xc0018f0180, 0x16}, {0xc000982978, 0x1, 0x1}, {0xc001885968, 0x3, 0x8}, {0x0, ...}, ...})
    /go/src/github.com/DataDog/datadog-agent/pkg/clusteragent/clusterchecks/dispatcher_main.go:149 +0xee
github.com/DataDog/datadog-agent/pkg/clusteragent/clusterchecks.(*dispatcher).Unschedule(0xc0023fb7f8?, {0xc002759600?, 0x1, 0x1?})
    /go/src/github.com/DataDog/datadog-agent/pkg/clusteragent/clusterchecks/dispatcher_main.go:120 +0x4e5
github.com/DataDog/datadog-agent/pkg/autodiscovery/scheduler.(*MetaScheduler).Unschedule(0xc000e14eb8, {0xc002759600?, 0x1, 0x1})
    /go/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/scheduler/meta.go:94 +0x131
github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).applyChanges(0xc0014977c0, {{0x0, 0x0, 0x0}, {0xc002759600, 0x1, 0x1}})
    /go/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:452 +0x3e5
github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).processDelService(0xc0014977c0, {0x3bd4230, 0xc001e1b080}, {0x3be8ac0?, 0xc0000cc7d0?})
    /go/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:427 +0x165
github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).serviceListening(0xc0014977c0)
    /go/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:114 +0x29a
created by github.com/DataDog/datadog-agent/pkg/autodiscovery.NewAutoConfig
    /go/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:69 +0x5b
```

**Describe what you expected:**
Agent to not crash.

**Steps to reproduce the issue:**
Unsure. Seems to have been happening sporadically for about a week now. Not sure what changed which might have started this. We've been running the same agent for about 2 weeks, but only seeing crashes start about a week ago.

Help in collecting additional information is would be useful.

**Additional environment details (Operating System, Cloud provider, etc):**
AWS EKS 1.23 & Bottlerocket OS

--- a/pkg/clusteragent/clusterchecks/dispatcher_configs.go
+++ b/pkg/clusteragent/clusterchecks/dispatcher_configs.go
@@ -90,12 +90,7 @@ func (d *dispatcher) removeConfig(digest string) {
 
 	node, found := d.store.getNodeStore(d.store.digestToNode[digest])
 
-	checkName := ""
-	if found {
-		node.RLock()
-		checkName = node.digestToConfig[digest].Name
-		node.RUnlock()
-	}
+	checkName := d.store.digestToConfig[digest].Name
 
 	delete(d.store.digestToNode, digest)
 	delete(d.store.digestToConfig, digest)

--- a/pkg/clusteragent/clusterchecks/dispatcher_configs.go
+++ b/pkg/clusteragent/clusterchecks/dispatcher_configs.go
@@ -101,18 +101,26 @@ func (d *dispatcher) removeConfig(digest string) {
 	delete(d.store.digestToConfig, digest)
 	delete(d.store.danglingConfigs, digest)
 
-	for k, v := range d.store.idToDigest {
-		if v == digest {
-			configsInfo.Delete(node.name, checkName, string(k), le.JoinLeaderValue)
-			delete(d.store.idToDigest, k)
+	// This is a list because each instance in a config has its own check ID and
+	// all of them need to be deleted.
+	var checkIDsToRemove []checkid.ID
+	for checkID, checkDigest := range d.store.idToDigest {
+		if checkDigest == digest {
+			checkIDsToRemove = append(checkIDsToRemove, checkID)
+			delete(d.store.idToDigest, checkID)
 		}
 	}
 
 	// Remove from node configs if assigned
 	if found {
 		node.Lock()
+		nodeName := node.name
 		node.removeConfig(digest)
 		node.Unlock()
+
+		for _, checkID := range checkIDsToRemove {
+			configsInfo.Delete(nodeName, checkName, string(checkID), le.JoinLeaderValue)
+		}
 	}
 }
 

--- a/pkg/clusteragent/clusterchecks/dispatcher_test.go
+++ b/pkg/clusteragent/clusterchecks/dispatcher_test.go
@@ -400,6 +400,36 @@ func TestDanglingConfig(t *testing.T) {
 	assert.Equal(t, 0, len(dispatcher.store.danglingConfigs))
 }
 
+func TestUnscheduleDanglingConfig(t *testing.T) {
+	testDispatcher := newDispatcher()
+
+	testConfig := integration.Config{
+		Name:         "cluster-check-example",
+		ClusterCheck: true,
+		Instances: []integration.Data{
+			// Define 2 to test that all of them are deleted and not just 1.
+			integration.Data("tags: [\"t1:v1\"]"),
+			integration.Data("tags: [\"t2:v2\"]"),
+		},
+	}
+
+	// False because because no node is available
+	assert.False(t, testDispatcher.shouldDispatchDanling())
+
+	// No nodes created, so it will not get assigned
+	testDispatcher.Schedule([]integration.Config{testConfig})
+
+	testDispatcher.Unschedule([]integration.Config{testConfig})
+
+	configs, err := testDispatcher.getAllConfigs()
+	require.NoError(t, err)
+	assert.Empty(t, configs)
+
+	// Also check that internal structures not used by getAllConfigs() are cleaned up
+	assert.Empty(t, testDispatcher.store.danglingConfigs)
+	assert.Empty(t, testDispatcher.store.idToDigest)
+}
+
 func TestReset(t *testing.T) {
 	dispatcher := newDispatcher()
 	config := generateIntegration("cluster-check")

--- a/releasenotes-dca/notes/fix-panic-dca-remove-dangling-config-6978bb2869634784.yaml
+++ b/releasenotes-dca/notes/fix-panic-dca-remove-dangling-config-6978bb2869634784.yaml
@@ -1,0 +1,12 @@
+# Each section from every release note are combined when the
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fixes a panic in the Cluster Agent that happens when trying to unschedule a
+    check that has not been dispatched to any runner.

